### ISSUE TITLE
Docker 1.11.0 daemon start failure after fresh install on CentOS 7

### ISSUE THREAD

**Issue by LPardue** [docker/docker#22204]
**Output of `docker version`:**

```
Client:
 Version:      1.11.0
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   4dc5990
 Built:        Wed Apr 13 18:40:36 2016
 OS/Arch:      linux/amd64
Cannot connect to the Docker daemon. Is the docker daemon running on this host?
```

**Output of `docker info`:**

```
Cannot connect to the Docker daemon. Is the docker daemon running on this host?
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Physical box running CentOS 7 with kernel 3.10.0-327.13.1.el7.x86_64. Docker installed via https://yum.dockerproject.org/

**Steps to reproduce the issue:**
1. Install docker on CentOS 7 following "Install with yum" instructions on https://docs.docker.com/engine/installation/linux/centos/
2. Attempt to start the docker daemon with `sudo systemctl start docker`
3. Failure to start the daemon

**Describe the results you received:**
`sudo systemctly start docker` returns with
`Job for docker.service failed because the control process exited with error code. See "systemctl status docker.service" and "journalctl -xe" for details.`

`sudo systemctl status docker -l` returns 

```
docker.service - Docker Application Container Engine
   Loaded: loaded (/usr/lib/systemd/system/docker.service; disabled; vendor preset: disabled)
   Active: failed (Result: exit-code) since Wed 2016-04-20 17:20:38 UTC; 5s ago
     Docs: https://docs.docker.com
  Process: 10321 ExecStart=/usr/bin/docker daemon -H fd:// (code=exited, status=1/FAILURE)
 Main PID: 10321 (code=exited, status=1/FAILURE)

Apr 20 17:20:37 [host redacted] docker[10321]: time="2016-04-20T17:20:37.890490133Z" level=warning msg="devmapper: Usage of loopback devices is strongly discouraged for production use. Please use `--storage-opt dm.thinpooldev` or use `man docker` to refer to dm.thinpooldev section."
Apr 20 17:20:37 [host redacted] docker[10321]: time="2016-04-20T17:20:37.915473736Z" level=info msg="devmapper: Creating filesystem xfs on device docker-253:0-134417280-base"
Apr 20 17:20:38 [host redacted] docker[10321]: time="2016-04-20T17:20:38.258365176Z" level=info msg="devmapper: Successfully created filesystem xfs on device docker-253:0-134417280-base"
Apr 20 17:20:38 [host redacted] docker[10321]: time="2016-04-20T17:20:38.299893813Z" level=info msg="Graph migration to content-addressability took 0.00 seconds"
Apr 20 17:20:38 [host redacted] docker[10321]: time="2016-04-20T17:20:38.314537929Z" level=info msg="Firewalld running: true"
Apr 20 17:20:38 [host redacted] docker[10321]: time="2016-04-20T17:20:38.501717532Z" level=fatal msg="Error starting daemon: Error initializing network controller: Error creating default \"bridge\" network: failed to parse pool request for address space \"LocalDefault\" pool \"\" subpool \"\": could not find an available predefined network"
Apr 20 17:20:38 [host redacted] systemd[1]: docker.service: main process exited, code=exited, status=1/FAILURE
Apr 20 17:20:38 [host redacted] systemd[1]: Failed to start Docker Application Container Engine.
Apr 20 17:20:38 [host redacted] systemd[1]: Unit docker.service entered failed state.
Apr 20 17:20:38 [host redacted] systemd[1]: docker.service failed.
```

**Describe the results you expected:**
I expected the service to start and allow me to verify the installation by running `sudo docker run hello-world`

**Additional information you deem important (e.g. issue happens only occasionally):**
This issue occurs 100% of the time. I have no legacy issues stemming from docker upgrades as this machine has never had docker running on it. I have seen other issues that sound similar to mine but none of the solutions (notably deleting `/var/lib/docker/network`) seem to allow the docker daemon to start.

**Comment by thaJeztah** [comment]
This error;

    Error starting daemon: Error initializing network controller: Error creating default \"bridge\" network: failed to parse pool request for address space \"LocalDefault\" pool \"\" subpool \"\": could not find an available predefined network"

Looks like a duplicate of https://github.com/docker/docker/issues/21170

**Comment by LPardue** [comment]
Thank you for the link @thaJeztah.

Running `route -n` reported a route similar problem route `172.16.0.0 ` with netmask `255.240.0.0`. If I delete this route, I can start the Docker service successfully. After it is started I see the route:

```
172.17.0.0      0.0.0.0         255.255.0.0     U     0      0        0 docker0
```

So while it seems like I have a fix, I do not understand the root cause behind the error. Would it be possible to explain this please?

**Comment by aboch** [comment]
@LPardue 
On a fresh install, docker looks for a subnet to assign to the default network `bridge` (backed by the linux bridge `docker0`). It goes through a list of predefined subnets and looks for an eligible one, which is one which does not overlaps with an existing route in your host.
If it cannot find any good subnet, it gives up as it would not be able to provide an operational default `bridge` network.

The predefined subnets are `172.[17-31].0.0/16` and  `192.168.[0-240].0/20`.

**Comment by LPardue** [comment]
@aboch Thank you for the information. I have been able to address the issue and run the docker daemon successfully.

The root cause of this issue was the conflict between `docker0` IP defaults and my server configuration. The solution to this problem was to supply an available specific IP and netmask for docker0 using the daemon option `--bip` as described at https://docs.docker.com/engine/userguide/networking/default_network/custom-docker0/. Removing the conflicting entries in the routing table was not a viable solution because they are actively by other system components.

I did not find the error message particularly useful in identifying the specific issue here, and I have not found documentation (via simple searches) that returns the predefined subnet details presented on this ticket. I think some improvement to this area would have allowed me to troubleshoot and fix unsupported.

**Comment by thaJeztah** [comment]
@LPardue looks like these messages/errors are generated by libnetwork (e.g. https://github.com/docker/docker/blob/3d13fddd2bc4d679f0eaa68b0be877e5a816ad53/vendor/src/github.com/docker/libnetwork/ipam/allocator.go#L415). Do you have suggestions for a different error message that would've helped you find a solution? (e.g. mentioning `--bip` in the error message?)

**Comment by LPardue** [comment]
I think there are a few aspects here:

> If it cannot find any good subnet, it gives up as it would not be able to provide an operational default bridge network.

I think this fail-fast action is sensible and gives ideal opportunity to report the real problem so that people can fix it. I am relatively new to Docker and the error message string didn't make total sense to me, breaking it down:

1. `Error starting daemon:` - OK, thats the thing I am trying to run
2. `Error initializing network controller:` - OK, its a network issue
3. `Error creating default \"bridge\" network:` - OK, the Docker introductory material explains the networking model
4. `failed to parse pool request for address space \"LocalDefault\" pool \"\" subpool \"\":` - I don't know what this means and a Google mainly just lists GitHub issues. These leads onto some of the  `rm -rf /var/lib/docker/network` advice that didn't help me. The code looks like it should print out three paramaters but that pool and subpool are empty. The empty values combined with the quote escaping makes it hard to parse (and enter into a search query).
5. `could not find an available predefined network` - I don't know what the predefined networks are (well I do now...).

So I think it would be helpful to list the networks that were attempted and failed. This could be explicit in the message or a link to that information elsewhere. The next sentence, like you suggest, could say something like `You can use the --bip option to assign the bridge an IP and netmask on an available (network|subnet).`

Possibly unrelated I spotted some typos at https://github.com/docker/docker/blob/3d13fddd2bc4d679f0eaa68b0be877e5a816ad53/vendor/src/github.com/docker/libnetwork/ipam/allocator.go#L384 

`no default pool availbale for non-default addresss spaces` should be `no default pool available for non-default address spaces`

**Comment by mavenugo** [comment]
@aboch i think there is an ask to change the error message ? Is this a candidate for closing ?

**Comment by aboch** [comment]
@mavenugo I now linked the libnetwork PR which will fix this. I incorrectly thought it was already mentioned. Nevertheless, the right pipeline would be "needs review", I wil lchange that. Thanks.
